website/blog/2025-03-19-20-must-try-mongodb-queries-operations-ferretdb.md

```diff
@@ -0,0 +1,437 @@
+
+```js
+db.books.aggregate([
+  {
+    $group: {
+      _id: '$genres',
+      average_rating: { $avg: '$analytics.average_rating' }
+    }
+  }
+])
+```
+
+### 15. Lookup (Join queries)
+
+Sometimes, you need to combine data from multiple collections – just like SQL joins in relational databases.
+This is useful when you store related data in different collections but want to retrieve them together in a single query.
+
+Suppose there's another collection `publishers` with more details on the publishers, as shown below:
+
+```js
+db.publishers.insertMany([
+  {
+    _id: 1,
+    name: 'T. Egerton',
+    location: 'London, United Kingdom',
+    established: 1780
+  },
+  {
+    _id: 2,
+    name: 'Harper & Brothers',
+    location: 'New York, United States',
+    established: 1817
+  },
+  {
+    _id: 3,
+    name: 'Lackington, Hughes, Harding, Mavor & Jones',
+    location: 'London, United Kingdom',
+    established: 1790
+  }
+])
+```
+
+You can join the data from both collections (`books` and `publishers`) using `$lookup`.
+
+```js
+db.books.aggregate([
+  {
+    $lookup: {
+      from: 'publishers',
+      localField: 'publication.publisher.name',
+      foreignField: 'name',
+      as: 'publisher_details'
+    }
+  }
+])
+```
+
+This should fetch the books and their corresponding publishers' details in a single query.
+
+## Geospatial queries
+
+### 16. Find books published in a specific location
+
+Some books aren't just about different places – they're from them.
+Geospatial queries help you query location-based data, like books published in a specific city or country.
+If a publisher's location is stored as GeoJSON points, FerretDB lets you query by geography instead of just text.
+
+What if you want to search for books published in a specific city.
+Instead of manually stating their locations, you can store and query precise geographic coordinates using `$geoWithin`:
+
+Using London's longitude and latitude (`[-0.1276, 51.5072]`), let's run some queries to find books published within a `1km` radius.
+Note that the distance is in radians, so we need to convert it to the Earth's radius.
+Since the Earth's radius is about 6378.1km, we divide `1km` by the Earth's radius to get the distance in radians.
+
+```js
+db.books.find({
+  'publication.publisher.location.geolocation': {
+    $geoWithin: {
+      $centerSphere: [[-0.1276, 51.5072], 1 / 6378.1]
+    }
+  }
+})
+```
+
+## Diagnostic operations
+
+### 17. Monitor operations with `currentOp`
+
+Sometimes, your database operations become slow, and you wonder what's causing the delay.
+`currentOp()` lets you inspect active operations, and find long-running queries that may overwhelm your instance.
+
+To actually see something in `currentOp()`, we need a long-running operation.
+Let's simulate a background operation by inserting 10,000 documents and running a `find` query that runs 1,000 times:
```

Comment:
By the way, the example is not a long-running operation. It's many operations.

Reply:
@Fashander